src/pdf_parser/pdf_utils/ocr.py

```diff
@@ -1,11 +1,329 @@
+
+    1. A block consists of a heading and a paragraph of text below it. In this case the block consists of 2
+    paragraphs, the first being the heading and the second being the text.
+    2. A block consists of a paragraph of
+    text. In this case the block and the paragraph have the same coordinates.
+    3. A block consists of multiple  paragraphs of text. For example, if we have a list of items,
+    each item is a paragraph.
+
+    Given this, and given that we want to cross-reference the returned layout with the one returned by
+    layoutparser + heuristics, we store blocks and paragraphs separately as "GoogleBlocks". To see the utility of this,
+    consider 2 cases:
+
+    1. Layout parser classifies a heading and a paragraph of text below it as 2 separate blocks,
+     one for a heading and another for the following paragraph. In this case, the best coordinate matches
+      with LayoutParser will be on what google classifies as paragraphs.
+    2. Layout parser detects a list of items as a single block but is unable to detect the individual items.
+    In this case, the best matches will be on what google classifies as blocks. So we want to merge the text
+    from the paragraphs into the blocks.
+
+    Regarding point 2, Google appears to be able to detect the individual items in a list more reliably than
+    layoutparser, a point we can explore in the future to get the best results by combining the two approaches.
+    For future usability, we also store block objects with paragraphs as sub-objects.
+
+    Args:
+        image: Image to extract document bounds from.
+
+    Returns:
+        List of GoogleBlocks, List of GoogleTextSegments, List of GoogleTextSegments, List of GoogleTextSegments
+    """
+
+    def _get_modal_string(string_list: List[str]) -> Optional[str]:
+        if len(string_list) > 0:
+            return max(set(block_languages), key=block_languages.count)
+        else:
+            return None
+
+    content = image_bytes(image)
+    client = vision.ImageAnnotatorClient()
+    image = types.Image(content=content)  # type: ignore
+
+    # TODO: Handle errors. Hit a 503.
+    response = client.document_text_detection(image=image)  # type: ignore
+    document = response.full_text_annotation
+
+    breaks = vision.enums.TextAnnotation.DetectedBreak.BreakType
+    lines = []
+    fully_structured_blocks = []
+    paragraph_text_segments = []
+    block_text_segments = []
+    for page in document.pages:
+        for block in page.blocks:
+            block_languages = []
+            default_dict = defaultdict(list)
+            for paragraph in block.paragraphs:
+                default_dict["paragraph_confidences"].append(paragraph.confidence)
+                para = ""
+                line = ""
+                para_languages = (
+                    []
+                )  # languages stored at word level. Detect then take mode.
+                for word in paragraph.words:
+                    try:
```

Comment:
I would split this up as it's a long block of code and thus hard to test and develop upon. 

For example if we had a function like "process_word" or something with a parameters, return type, docstring then we wouldn't have to comment the code and could make unit tests, whereas right now if we want to test how we process a word we need to call extract_google_layout.

Reply:
Parking for now